src/Common/src/Interop/Windows/advapi32/Interop.GetServiceDisplayName.cs

```diff
@@ -0,0 +1,19 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Win32.SafeHandles;
+using System;
+using System.Buffers;
+using System.ComponentModel;
+using System.Runtime.InteropServices;
+using System.Text;
+
+internal partial class Interop
+{
+    internal partial class Advapi32
+    {
+        [DllImport(Libraries.Advapi32, EntryPoint = "GetServiceDisplayNameW", CharSet = System.Runtime.InteropServices.CharSet.Unicode, SetLastError = true)]
+        internal static extern bool GetServiceDisplayName(SafeServiceHandle SCMHandle, string serviceName, ref char displayName, ref int displayNameLength);
```

Comment:
I think it is better to use `char*` for string pointers in PInvoke signatures. `ref char` obfuscates the intent. It is hard to see that it is actually pointer to string.

Reply:
We are not consistent with this.

Reply:
cc @JeremyKuhne - we should probably be consistent.

Reply:
I think we had settled on not using ref char instead of char* to potentially avoid an extra frame in addition to clarity. In this case we're stuck with extra frames no matter what due to the safe handle.

On a side note: one weird gotcha we have in this space is that pinning an empty array behaves differently than pinning an empty span.  

``` C#
static unsafe void PinMe()
{
    int[] emptyArray = new int[0];
    Span<int> emptySpan = new Span<int>(emptyArray);
    fixed (int* ea = emptyArray)
    fixed (int* es = emptySpan)
    fixed (int* esr = &MemoryMarshal.GetReference(emptySpan))
    {
        Console.WriteLine($"Array: 0x{(ulong)ea:X}, Span: 0x{(ulong)es:X}, Address of ref span: 0x{(ulong)esr:X}");
    }
}
```

Yields something like:

```
Array: 0x0, Span: 0x0, Address of ref span: 0x27D97CD1588
```

If the source array actually had length, these would all return the same pointer. This actually generated regressions in System.Drawing as GDI+ cared about the pointer in APIs when we sent the pointer and a length of zero (user passed us no points and we passed no points along to GDI+).

Reply:
> pinning an empty array behaves differently than pinning an empty span.

Your example shows that pinning empty array works same way as pinning empty span.

It is only when you use the low-level `MemoryMarshal.GetReference` API that you will see the difference. We have a number of places in CoreFX that use `MemoryMarshal.GetReference` to pin Span that were introduced before we have proper support for pinning Span. We should move away from using `MemoryMarshal.GetReference` for pinning Span.

Reply:
> Your example shows that pinning empty array works same way as pinning empty span.

Yes, I should have been clearer. `&MemoryMarshal.GetReference` was the only "pinning" option prior to C# 7.3. I'll open a tracking issue to pull usages.

Reply:
@jkotas Incidentally, does using `ref char` instead of `char*` have the same impact with the marshaller as `MemoryMarshal.GetReference`? I'll validate when I get some time.

Reply:
Should we make a breaking change to `MemoryMarshal.GetReference` to change to the ideal behavior? I imagine not much code has been written to use it on spans yet.

Reply:
`MemoryMarshal.GetReference` was designed to give you the Span pointer with zero overhead and without any normalization.

If we added normalization to it, it would make it impossible to write low-level high perf code that works on Span. I am pretty sure number of our micro-benchmarks would regress.